pkg/telemetry/statsworker.go

```diff
@@ -15,158 +15,178 @@ type StatsWorker struct {
+		analyticsStat := coalesce(analyticsStats)
+		if analyticsStat == nil {
+			continue
 		}
+
+		s.patch(analyticsStat, ts, trackID, livekit.StreamType_UPSTREAM)
+		stats = append(stats, analyticsStat)
 	}
+	s.incomingPerTrack = make(map[livekit.TrackID][]*livekit.AnalyticsStat, 0)
+
 	return stats
 }
 
-func (s *StatsWorker) update(stats *Stats, ts *timestamppb.Timestamp) *livekit.AnalyticsStat {
-	if stats.totalBytes == 0 {
+func (s *StatsWorker) patch(
+	analyticsStat *livekit.AnalyticsStat,
+	ts *timestamppb.Timestamp,
+	trackID livekit.TrackID,
+	kind livekit.StreamType,
+) {
+	analyticsStat.TimeStamp = ts
+	analyticsStat.TrackId = string(trackID)
+	analyticsStat.Kind = kind
+	analyticsStat.RoomId = string(s.roomID)
+	analyticsStat.ParticipantId = string(s.participantID)
+	analyticsStat.RoomName = string(s.roomName)
+}
+
+func (s *StatsWorker) Close() {
+	s.Update()
+}
+
+func coalesce(stats []*livekit.AnalyticsStat) *livekit.AnalyticsStat {
+	if len(stats) == 0 {
 		return nil
 	}
 
-	next := stats.next
-	stats.next = &livekit.AnalyticsStat{
-		Kind:          next.Kind,
-		RoomId:        string(s.roomID),
-		ParticipantId: string(s.participantID),
-		RoomName:      string(s.roomName),
+	// average score of all available stats
+	score := float32(0.0)
+	for _, stat := range stats {
+		score += stat.Score
+	}
+	score = score / float32(len(stats))
+
+	// aggregate streams across all stats
+	maxRTT := make(map[uint32]uint32)
+	maxJitter := make(map[uint32]uint32)
+	analyticsStreams := make(map[uint32]*livekit.AnalyticsStream)
+	for _, stat := range stats {
+		//
+		// For each stream (identified by SSRC) consolidate reports.
+		// For cumulative stats, take the latest report.
+		// For instantaneous stats, take maximum (or some other appropriate representation)
+		//
+		for _, stream := range stat.Streams {
+			ssrc := stream.Ssrc
+			analyticsStream := analyticsStreams[ssrc]
+			if analyticsStream == nil {
+				analyticsStreams[ssrc] = stream
+				maxRTT[ssrc] = stream.Rtt
+				maxJitter[ssrc] = stream.Jitter
+				continue
+			}
+
+			if stream.TotalPrimaryPackets <= analyticsStream.TotalPrimaryPackets {
+				// total count should be monotonically increasing
+				continue
+			}
+
+			analyticsStreams[ssrc] = stream
+			if stream.Rtt > maxRTT[ssrc] {
+				maxRTT[ssrc] = stream.Rtt
+			}
+
+			if stream.Jitter > maxJitter[ssrc] {
+				maxJitter[ssrc] = stream.Jitter
+			}
+		}
 	}
 
-	next.TimeStamp = ts
-	next.TotalPackets = uint64(stats.totalPackets - stats.prevPackets)
-	next.TotalBytes = stats.totalBytes - stats.prevBytes
-	next.PacketLost = stats.totalPacketsLost - stats.prevPacketsLost
-	next.ConnectionScore = stats.connectionScore
-
-	stats.prevPackets = stats.totalPackets
-	stats.prevBytes = stats.totalBytes
-	stats.prevPacketsLost = stats.totalPacketsLost
-	return next
-}
+	streams := make([]*livekit.AnalyticsStream, 0, len(analyticsStreams))
+	for ssrc, analyticsStream := range analyticsStreams {
+		stream := &livekit.AnalyticsStream{
```

Comment:
I wanted to copy the structure like `stream := *analyticsStream` and change only a couple of fields necessary. But, `vim-go` warns about locks getting copied. I guess is a protobuf thing. Was looking for a Clone/Copy method of the protobuf generated structure. Looks like there is nothing there. So, ended up with this blown up copy.

Reply:
can use protobuf.Clone

Reply:
Not able to use it. It does not allow me to modify fields in the cloned message.

Reply:
why it does not allowed, does this code can not work?
```
stream := proto.Clone(analyticStream).(*liveit.AnalyticsStream) 
stream.Ssrc = ssrc 
...

Reply:
That worked @cnderrauber . Thank you. Forgot to do the casting.